packages/web3torrent/src/library/web3torrent-lib.ts

```diff
@@ -124,92 +138,143 @@ export default class WebTorrentPaidStreamingClient extends WebTorrent {
+    log(
+      `newSeederBalance: ${newSeederBalance} payment: ${payment}, funds for peer: ${this.peersList[infoHash][peerId].funds}`
+    );
   }
 
-  protected transferFunds(wire: PaidStreamingWire) {
-    // [ George ] I think this is where the leecher could call channelClient.updateChannel(). A first iteration might just do this without any UI or checks.
+  protected async transferFunds(wire: PaidStreamingWire) {
+    const channelId = Object.keys(this.channelClient.openChannels)[0]; // TODO use proper index to get correct channel (inspect some lookup from wire to channelId?)
+
+    // (window.channelProvider as FakeChannelProvider).playerIndex = 1;
 
-    // INFO: Assumed payment. This could emit an event on the UI to ask for more funds, or be automatic. Dunno.
-    setTimeout(() => {
-      wire.paidStreamingExtension.payment('50');
-    }, 500);
+    await this.channelClient.makePayment(channelId, bigNumberify(REQUEST_RATE.mul(10)).toString());
+    const newSeederBalance = bigNumberify(this.channelClient.openChannels[channelId].seederBalance);
+    log(`payment made for channel ${channelId}, newSeederBalance: ${newSeederBalance}`);
   }
 
   protected setupTorrent(torrent: PaidStreamingTorrent) {
     if (torrent.usingPaidStreaming) {
       return torrent;
     }
     torrent.on('infoHash', () => {
-      this.allowedPeers = {...this.allowedPeers, [torrent.infoHash]: {}};
+      this.peersList = {...this.peersList, [torrent.infoHash]: {}};
     });
     torrent.on(TorrentEvents.WIRE, (wire: PaidStreamingWire) => {
       this.setupWire(torrent, wire);
     });
 
-    torrent.on(TorrentEvents.NOTICE, (wire, {command, data}) => {
+    torrent.on(TorrentEvents.NOTICE, async (wire, {command, data}) => {
       log(`< ${command} received from ${wire.peerExtendedHandshake.pseAccount}`, data);
+      let turnNum: number;
+      let channelId: string;
       switch (command) {
         case PaidStreamingExtensionNotices.STOP:
           wire.paidStreamingExtension.ack();
           wire.choke();
-          this.transferFunds(wire);
+          if (!torrent.done) {
+            await this.transferFunds(wire);
+          }
           break;
         case PaidStreamingExtensionNotices.START:
-          // [ George ] Here we can call channelClient.createChannel()
           wire.paidStreamingExtension.ack();
           this.jumpStart(torrent, wire);
           break;
-        case PaidStreamingExtensionNotices.PAYMENT:
-          this.loadFunds(torrent.infoHash, wire.peerExtendedHandshake.pseAccount, data.hash);
-          this.unblockPeer(torrent.infoHash, wire, wire.peerExtendedHandshake.pseAccount);
+        case PaidStreamingExtensionNotices.MESSAGE:
+          await this.channelClient.pushMessage(JSON.parse(data.message));
+          turnNum = Number(JSON.parse(data.message).data.turnNum);
+          channelId = JSON.parse(data.message).data.channelId;
+          if (
+            JSON.parse(data.message).recipient === this.pseAccount &&
```

Comment:
I think I read something about the turns that you guys use to handle payment channels. But is it relevant here?
In any case, couldn't this be something handled inside the channelClient? There's a lot of what it seems like channel-related logic here.

Reply:
Yes I think we can eventually hide a lot of this in the channel client.